akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -498,72 +487,94 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
+  // FIXME: Compression table must be owned by each channel instead of having a global one
+  private def createInboundCompressionTable(): LiteralCompression = {
+    val advertActorRef = (ref: String, id: Int) ⇒ {
+      val remoteAddress = ActorPath.fromString(ref).address // TODO expensive
+      val addr = localAddress.address
+      log.error(s"advertise compression to [$remoteAddress], of [$ref] => [$id] ")
+      sendControl(remoteAddress, CompressionProtocol.ActorRefCompressionAdvertisement(addr, ref, id))
+    }
+    val dontAdvertise: (Any, Any) ⇒ Unit = (a, i) ⇒ {
+      ()
+    }
+    if (remoteSettings.ArteryCompressionSettings.enabled) new ActualCompressionTables(system, advertActorRef, dontAdvertise)
+    else NoLiteralCompression
+  }
+
+  def createEncoder(pool: EnvelopeBufferPool, compression: LiteralCompression): Flow[Send, EnvelopeBuffer, NotUsed] =
+    Flow.fromGraph(new Encoder(localAddress, system, compression, pool)) // TODO pass in
 
-  def encoder: Flow[Send, EnvelopeBuffer, NotUsed] = createEncoder(envelopePool)
+  def encoder(compression: LiteralCompression): Flow[Send, EnvelopeBuffer, NotUsed] = createEncoder(envelopePool, compression)
 
   val messageDispatcherSink: Sink[InboundEnvelope, Future[Done]] = Sink.foreach[InboundEnvelope] { m ⇒
     messageDispatcher.dispatch(m.recipient, m.recipientAddress, m.message, m.senderOption)
   }
 
-  def createDecoder(pool: EnvelopeBufferPool): Flow[EnvelopeBuffer, InboundEnvelope, NotUsed] = {
+  def createDecoder(pool: EnvelopeBufferPool, compression: LiteralCompression): Flow[EnvelopeBuffer, InboundEnvelope, NotUsed] = {
     val resolveActorRefWithLocalAddress: String ⇒ InternalActorRef =
       recipient ⇒ provider.resolveActorRefWithLocalAddress(recipient, localAddress.address)
-    Flow.fromGraph(new Decoder(localAddress, system, resolveActorRefWithLocalAddress, compression, pool))
+    Flow.fromGraph(new Decoder(localAddress, system, resolveActorRefWithLocalAddress, compression, pool)) // TODO pass-in
   }
 
-  def decoder: Flow[EnvelopeBuffer, InboundEnvelope, NotUsed] = createDecoder(envelopePool)
+  def decoder(compression: LiteralCompression): Flow[EnvelopeBuffer, InboundEnvelope, NotUsed] = createDecoder(envelopePool, compression)
 
   def inboundSink: Sink[InboundEnvelope, NotUsed] =
     Flow[InboundEnvelope]
       .via(new InboundHandshake(this, inControlStream = false))
       .via(new InboundQuarantineCheck(this))
       .to(messageDispatcherSink)
 
-  def inboundFlow: Flow[EnvelopeBuffer, ByteString, NotUsed] = {
+  def inboundFlow(compression: LiteralCompression): Flow[EnvelopeBuffer, ByteString, NotUsed] = {
     Flow.fromSinkAndSource(
-      decoder.to(inboundSink),
+      decoder(compression).to(inboundSink),
       Source.maybe[ByteString].via(killSwitch.flow))
   }
 
   def inboundLargeFlow: Flow[EnvelopeBuffer, ByteString, NotUsed] = {
+    // TODO large flow does not care about compression IMO. right?  
```

Comment:
Good point @patriknw.
So if we'd share the same table in this and the other flow then it would have to guard from concurrent access I think... Will think about solving this (2 tables also introduces some questions)